### ISSUE TITLE
Error with MongoDB - DatabaseWatcher Using change streams

### ISSUE THREAD

**Issue by scrivna** [RocketChat/Rocket.Chat#26615]
### Description:

While running the upgrade from 4.8.1 to 5.0.3 I receive a MongoDB error message related to watching the database for changes.

### Steps to reproduce:

1. Using hosted MongoDB Atlas 3 node replica set with version 5.0.10
2. Upgrade process begins on server then crashes with the following error...

### Expected behavior:

Working instance

### Actual behavior:

Crashes during startup

### Server Setup Information:

- Version of Rocket.Chat Server: 
- Operating System: Ubuntu
- Deployment Method: docker
- Number of Running Instances: 1
- DB Replicaset Oplog: Yes
- NodeJS Version: 4.19.3
- MongoDB Version: 5.0.10

### Additional context

4.8.1 Was working fine.

I see there's been a recent change to how this works... from @sampaiodiego - https://github.com/RocketChat/Rocket.Chat/pull/26336

### Relevant logs:

```
[DatabaseWatcher] Using change streams
events.js:377
      throw er; // Unhandled 'error' event
      ^
MongoServerError: Error validating $match value for change streams. err=ns field has bson.D value that is not string or valid MongoDb RegEx: Error parsing value [{$in [rocketchat_message users rocketchat_subscription rocketchat_livechat_inquiry rocketchat_livechat_department_agents usersSessions rocketchat_permissions rocketchat_roles rocketchat_room meteor_accounts_loginServiceConfiguration instances rocketchat_integration_history rocketchat_integrations rocketchat_email_inbox pbx_events rocketchat_settings]}] to RegEx: Must specify $regex field
    at MessageStream.messageHandler (/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/cmap/connection.js:462:30)
    at MessageStream.emit (events.js:400:28)
    at MessageStream.emit (domain.js:475:12)
    at processIncomingData (/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/cmap/message_stream.js:108:16)
    at MessageStream._write (/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/cmap/message_stream.js:28:9)
    at writeOrBuffer (internal/streams/writable.js:358:12)
    at MessageStream.Writable.write (internal/streams/writable.js:303:10)
    at TLSSocket.ondata (internal/streams/readable.js:731:22)
    at TLSSocket.emit (events.js:400:28)
    at TLSSocket.emit (domain.js:475:12)
    at addChunk (internal/streams/readable.js:293:12)
    at readableAddChunk (internal/streams/readable.js:267:9)
    at TLSSocket.Readable.push (internal/streams/readable.js:206:10)
    at TLSWrap.onStreamRead (internal/stream_base_commons.js:188:23)
    at TLSWrap.callbackTrampoline (internal/async_hooks.js:130:17)
Emitted 'error' event on ChangeStream instance at:
    at closeWithError (/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/change_stream.js:355:22)
    at processError (/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/change_stream.js:448:12)
    at Readable.<anonymous> (/app/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb/lib/change_stream.js:364:33)
    at Readable.emit (events.js:400:28)
    at Readable.emit (domain.js:475:12)
    at emitErrorNT (internal/streams/destroy.js:106:8)
    at emitErrorCloseNT (internal/streams/destroy.js:74:3)
    at processTicksAndRejections (internal/process/task_queues.js:82:21) {
  ok: 0,
  code: 8000,
  codeName: 'AtlasError'
}
```

**Comment by sampaiodiego** [comment]
I'll take a look on this asap

**Comment by sampaiodiego** [comment]
@scrivna do you have a support contact on Atlas to ask what more details about the issue? you can attach the following code snippet, this is the operation throwing the error:
```javascript
const changeStream = db.watch([
  {
    "$match": {
      "operationType": {
        "$in": [
          "insert",
          "update",
          "delete"
        ]
      },
      "ns.coll": {
        "$in": [
          "rocketchat_message",
          "users",
          "rocketchat_subscription",
          "rocketchat_livechat_inquiry",
          "rocketchat_livechat_department_agents",
          "usersSessions",
          "rocketchat_permissions",
          "rocketchat_roles",
          "rocketchat_room",
          "meteor_accounts_loginServiceConfiguration",
          "instances",
          "rocketchat_integration_history",
          "rocketchat_integrations",
          "rocketchat_email_inbox",
          "pbx_events",
          "rocketchat_settings"
        ]
      }
    }
  }
])
```

meanwhile, you can tell rocket.chat to use Oplog instead of Change Streams. to make so, you'll have to make sure the env var `MONGO_OPLOG_URL` is being provided.. additionally to that, you should set `IGNORE_CHANGE_STREAM=true` .

**Comment by scrivna** [comment]
@sampaiodiego thanks for looking in to this, yes I have a contact at Atlas I can reach out to. Do you think this is something custom they're doing with their MongoDB installs? RC runs fine for me when i run it locally with the RC dev environment version of Mongo - I'm guessing it does for you as well.

**Comment by sampaiodiego** [comment]
> Do you think this is something custom they're doing with their MongoDB installs?

@scrivna yes, exactly.. the error even says `AtlasError`, it may be some limitation to avoid abusing, idk.. I really would like to understand what exactly is the issue so we can properly support Atlas.

we have rocket.chat 5.0.x running in multiple mongodb versions, so it is definitely supported.

**Comment by scrivna** [comment]
@sampaiodiego I've been trying to get hold of someone at MongoDB Atlas to help us with this. I'll keep you updated.

**Comment by scrivna** [comment]
@sampaiodiego I've heard back from the MongoDB team...

"This appears to be a bug with the change stream proxy layer. Our engineers have filed an internal ticket and it is being reviewed."

**Comment by sampaiodiego** [comment]
any news from them?

**Comment by sampaiodiego** [comment]
oh, you can opt-out on using change streams.. just make sure to provide the following env vars to rocket.chat:
```
MONGO_OPLOG_URL=mongodb://user:pass@atlas.address/local?replicaSet=rsName
IGNORE_CHANGE_STREAM=true
```

**Comment by scrivna** [comment]
Thanks for providing that en var - I tried this out and managed to get the 5.x branch running... however it came with a lot of issues, I think the biggest one was that when settings were updated they went to the DB but didn't update in memory, so I had to keep restarting the server to get it to pick up the updated settings, so sounds like a memory cache or something, also new messages wouldn't display unless the browser was refreshed, it kinda killed all the real-time features.

Anyway, I got a response back from Atlas "Just wanted to inform you that the bug has been resolved now." - and yes, I removed the IGNORE_CHANGE_STREAM flag and I have a 5.x branch running as standard now and everything seems to be working as normal now, now weird settings or message bugs visible.

Thanks for your help with this issue!
